### DISCUSSION TITLE
Turn off onBeforeRender in nested route

### DISCUSSION THREAD

**Discussion by dajinchu** [vikejs/vike#1075] [Discussion] [answered]
if `pages/app/+config` defines an onBeforeRender, but I want `pages/app/star-wars/+config` not to have any onBeforeRender at all (and for it to skip making network request to index.pageContext.json) how would I do that?

Do I just have to un-nest star-wars route in the file structure? Would be nice to be able to set `false` or something

**Comment by brillout** [comment]
I agree that'd be nice. In the meantime I believe setting a `onBeforeRender()` that does nothing should do the trick. Let me know whether that works for you.

**Reply by dajinchu** [reply]
Trouble for me is I need to not hit the server (usecase/reason below). Ever since the pageContextInitHasClientData stuff was introduced, when onbeforeRender is defined it will hit the server, even if it is client only (at least for my case)

My use case is that I'm overriding the back/forward button behavior by using onBeforeRoute. This is to handle Turbolinks' caching functionality (it takes a snapshot of the page before navigating and restores on back button). Weird use case, but it means that the routing is different on client vs server in order to route to a special page on back button to render from the snapshot. 

https://github.com/Alignable/bifrost/blob/master/bifrost/renderer/onBeforeRoute.ts


Ultimately just being able to turn off onBeforeRender would solve this but open to other ideas.

**Reply by brillout** [reply]
Have you tried to use [`meta` to change `onBeforeRender()`'s env](https://vite-plugin-ssr.com/migration/v1-design#onbeforerender-in-page-js)? Also note https://github.com/brillout/vite-plugin-ssr/issues/1043, so make sure you're using the latest version.

**Reply by brillout** [reply]
In principle, if you set the fake `onBeforeRender()` to `client-only` then this should "turn off" `onBeforeRender()`.

> Ever since the pageContextInitHasClientData stuff was introduced, when onbeforeRender is defined it will hit the server, even if it is client only (at least for my case)

That isn't how `pageContextInitHasClientData` is determined: the `pageContextInitHasClientData` value is `true` when you define some `pageContextInit` prop that is included in `passToClient`.

This means that, if there isn't a `onBeforeRender()` that is `server-only` and there isn't any `pageContextInit` prop included in `passToClient`, then client-side routing shouldn't make a request to your server.

**Reply by dajinchu** [reply]
Sorry for being unclear. We are on 0.4.135 and are using `client-only`. We do pass pageContextInit props to passToClient. We proxy requests to legacy backend at the fastify level, so we have to pass the proxy contents down via pageContextInit. 

I tried setting our "back button page" to have a different passToClient (empty array), since it does not need any pageContext, but I don't think `pageContextInitHasClientData` supports different passToClients?

Regardless, I have it working now by renaming props from pageContextInit inside the server-side `onBeforeRender` so that we aren't passing any `passToClient` keys through pageContextInit.

**Reply by brillout** [reply]
Replied [here](https://github.com/brillout/vite-plugin-ssr/discussions/1075#discussioncomment-6767044).

**Comment by brillout** [comment]
> I tried setting our "back button page" to have a different passToClient (empty array), since it does not need any pageContext, but I don't think `pageContextInitHasClientData` supports different passToClients?

The `pageContextInitHasClientData` value is determined based on the *previous* fetched `pageContext`. There isn't a way to predict whether the next `pageContext` will contain `pageContextInit` data.

I will have a look at supporting preventing global hooks from applying to some pages by allow these pages to set the hook value to `null`. (It may actually already work, I've to look.)

Let me know if you believe there is anything vite-plugin-ssr can do to reduce the amount of duct taping.

Thanks for reaching out and, as always, let me know if you run into issues.

**Reply by brillout** [reply, marked as answer]
> I will have a look at supporting preventing global hooks from applying to some pages by allow these pages to set the hook value to `null`.

Done, see 11202d2. I'll release in a couple of days. I've pre-released it as `0.4.136-commit-11202d2` if you don't want to wait.

**Reply by dajinchu** [reply]
works great, thanks! `Config` type does not allow null though (with strict mode on).

**Reply by brillout** [reply]
Thanks & done (fefc63ead5959aec6e6344f53b049d9cafeb2139).

**Reply by brillout** [reply]
Released in `0.4.138`.

**Reply by dajinchu** [reply]
Awesome thanks!
